SYS_ATL/new_eff.py

```diff
@@ -1262,6 +1273,290 @@ def bds(x,hi):
+
+    no_bound_change = (
+        AForAll([i], AImplies( AMay(bds(i,hi)),
+            AAnd(Commutes(a_bd, a1), Commutes(a_bd, a2))  )))
+    stmts_commute = (
+        AForAll([i,j], AImplies( AMay(AAnd( bds(i,hi), bds(j,hi),
+                                            AInt(i) < AInt(j) )),
+                                 AAnd(Commutes(a1_j, a2),
+                                      AllocCommutes(a1, a2)) )))
+
+    pred    = G(AAnd(no_bound_change, stmts_commute))
+    is_ok   = slv.verify(pred)
+    slv.pop()
+    if not is_ok:
+        raise SchedulingError(
+            f"Cannot fission loop over {i} at {loop.srcinfo}.")
+
+
+
+
+def Check_DeleteConfigWrite(proc, stmts):
+    assert len(stmts) > 0
+    ctxt = ContextExtraction(proc, stmts)
 
+    p       = ctxt.get_control_predicate()
+    G       = ctxt.get_pre_globenv()
+    ap      = ctxt.get_posteffs()
+    a       = G(stmts_effs(stmts))
+    stmtsG  = globenv(stmts)
+
+    slv     = SMTSolver(verbose=False)
+    slv.push()
+    a       = [E.Guard(AMay(p), a)]
+
+    # extract effects
+    WrG, Mod    = getsets([ES.WRITE_G, ES.MODIFY], a)
+    WrGp, RdGp  = getsets([ES.WRITE_G, ES.READ_G], ap)
+
+    # check that `stmts` does not modify any non-global data
+    only_mod_glob   = ADef(is_empty(LDiff(Mod,WrG)))
+    is_ok           = slv.verify(only_mod_glob)
+    if not is_ok:
+        slv.pop()
+        raise SchedulingError(
+            f"Cannot delete or insert statements at {stmts[0].srcinfo} "
+            f"because they may modify non-configuration data")
+
+    # get the set of config variables potentially modified by
+    # the statement block being focused on.  Filter out any
+    # such configuration variables whose values are definitely unchanged
+    def is_cfg_unmod_by_stmts(pt):
+        pt_e    = ABool(pt.name) if pt.typ == T.bool else AInt(pt.name) 
+        #cfg_unwritten = ADef( ANot(is_elem(pt, WrG)) )
+        cfg_unchanged = ADef( G(AEq(pt_e, stmtsG(pt_e))) )
+        return slv.verify(cfg_unchanged)
+    cfg_mod = set( pt.name for pt in get_point_exprs(WrG)
+                           if not is_cfg_unmod_by_stmts(pt) )
+
+    # consider every global that might be modified
+    cfg_mod_visible = set()
+    for pt in get_point_exprs(WrG):
+        pt_e    = ABool(pt.name) if pt.typ == T.bool else AInt(pt.name) 
+        is_written      = is_elem(pt, WrG)
+        is_unchanged    = G(AEq(pt_e, stmtsG(pt_e)))
+        is_read_post    = is_elem(pt, RdGp)
+        is_overwritten  = is_elem(pt, WrGp)
+
+        # if the value of the global might be read,
+        # then it must not have been changed.
+        safe_write  = AImplies( AMay(is_read_post), ADef(is_unchanged) )
+        if not slv.verify(safe_write):
+            # for s in stmts:
+            #     print(s)
+            # print('####')
+            # print(safe_write)
+            # print('***')
+            # print(G)
+            # print('***')
+            # print(stmtsG)
```

Comment:
I'm not going to mark every instance of this... please clean up the commented print debugs.